Add background refresh stat safeguards

samsymons opened the pull request:
Task/Issue URL: https://app.asana.com/0/1199333091098016/1199644311779247
CC: @bwaresiak 

**Description**:

This PR introduces a mechanism which checks whether a background task's completion handler has been called previously, and corrects the background refresh metrics if so.

**Steps to test this PR**:
1. Test the background refresh flow on a device through Xcode and verify that the typical flows work as expected; there are instructions inside `AppConfigurationFetch` for manually triggering background tasks

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPad

**OS Testing**:

* [ ] iOS 11
* [ ] iOS 12
* [ ] iOS 13
* [ ] iOS 14

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

samsymons committed:
Add checks for the background refresh completion status.

samsymons committed:
Test the app config fetch completion handler.

samsymons committed:
Remove a now-unneeded SwiftLint warning.

samsymons committed:
Refactor the config fetch tests.

samsymons committed:
Add an extra test to ensure that the count can’t be negative.

samsymons committed:
Avoid usage of `inout`.

samsymons committed:
Clean a SwiftLint warning.

samsymons committed:
Fix unit tests on < iOS 13 simulators.

samsymons committed:
Send changes to `lastCompletionStatus` through the main thread.

The expiration handler and background refresh task are run on different threads, so they need to use a shared queue to write to shared state.

Asavarkhul commented on DuckDuckGo/AppConfigurationFetch.swift:
```diff
@@ -95,6 +94,16 @@ class AppConfigurationFetch {
+        case newData
+
+        var success: Bool {
+            self == .noData || self == .newData
```
Could be simplified by `self != .expired` wdyt?

samsymons replied:
Yeah that's a fair suggestion, I wrote it this way originally in order to explicitly state the expected cases, to avoid the situation where new cases added in the future won't impact the logic – let's say a new `.error` case is added, the current implementation would still work as intended, whereas only checking for `.expired` would mean `success` returns true even for an `.error`.

All that being said, this enum is so simple that checking for `expired` makes sense. 🙂

Asavarkhul replied:
Ho yes I agree with your point of view 😏
The only dangerous stuff with computed properties like this is when there is no unit test on it, which could involve one day the typical issue your are mentioning ; So maybe just unit test it? 🤷‍♂️

Asavarkhul commented on DuckDuckGo/AppConfigurationFetch.swift:
```diff
@@ -313,3 +300,78 @@ class AppConfigurationFetch {
+        }
+
+        queue.async {
+            let newData = configurationFetcher.fetchConfigurationFiles(isBackground: true)
```
Since it's a Bool which is returned, why don't naming it like `let newDataExists = ...` ?
Wdyt?

samsymons replied:
Good suggestion!

samsymons committed:
Rename `newData` to `fetchedNewData`.

This reflects that an action took place, rather than `newData` which sounds like it contains a new object.

samsymons committed:
Simplify the `success` computed var on `BackgroundRefreshCompletionStatus`.

samsymons committed:
Include the expiration count as a parameter.

samsymons committed:
Simplify the background refresh completion handler.

bwaresiak commented on DuckDuckGo/AppConfigurationFetch.swift:
```diff
@@ -68,9 +70,6 @@ class AppConfigurationFetch {
     static private var lastConfigurationRefreshDate: Date
 
-    @UserDefaultsWrapper(key: .backgroundFetchTaskExpirationCount, defaultValue: 0)
-    static private var backgroundFetchTaskExpirationCount: Int
```
Shouldn't we also remove `backgroundFetchTaskExpirationCount`  Key? This seems to be defined in two places now - as an enum in `UserDefaultsWrapper.Key` and as a static let in `AppUserDefaults.Keys`? Also maybe we should be reusing the old string key as sort of migration?

bwaresiak reviewed (approved):
Great job @samsymons, one small comment inline. :)

samsymons committed:
Clean up some changes related to moving user defaults keys around.

samsymons committed:
Test that the BackgroundRefreshCompletionStatus enum is correct.
